### PR TITLE
Unittests, linting and code formatting using nox

### LINKED ISSUE #10: Add linting and code formatting
@salmannotkhan kindly identified that we had imported and unused the `sys` library. This could have been automagically fixed prior to committing if we had been using _code linting_ in a more automated workflow, e.g. using [flake8](https://flake8.pycqa.org/en/latest/) to follow PEP guidelines prior to commits. At the same time, we could also add _code formatting_ with [black](https://github.com/psf/black), and make both runnable via [nox](https://nox.thea.codes/en/stable/), which we use for running tests in #9. 

## Changes Needed

1. Add `flake8` and a range of sensible plugins, e.g. `flake8-import-order`, `flake8-builtins`, `flake8-quotes`, `flake8-docstrings`, and `flake8-black`. This requires adding flake8 as a `--dev` dependency via poetry. 
2. Add and configure `black` for code formatting.
3. Add nox to run flake8 and black as it will be used for running tests.
4. Add [pre-commit](https://pre-commit.com/) as a dev dependency and configure its TOML file accordingly to automate the above prior to a git commit.

See [here](https://ljvmiranda921.github.io/notebook/2018/06/21/precommits-using-black-and-flake8/) and [here](https://cjolowicz.github.io/posts/hypermodern-python-03-linting/) for examples of implementing the above.

### PR DESCRIPTION
**Pull request by davidverweij**
## Closes #10 .
And works on point 1 and 2 of #23 

## What Changed (and why)?
I've followed guide 2 ([Testing](https://cjolowicz.github.io/posts/hypermodern-python-02-testing/)) and 3 ([Linting](https://cjolowicz.github.io/posts/hypermodern-python-03-linting/)) from Claudio Jolowicz's Hypermodern Python article to implement @jawrainey's suggested code linting, unittesting and pre-commit hooks. More specifically, I've implemented the following:

  1. Set up testing using [`pytest`](https://docs.pytest.org/en/latest/) and [`coverage`](https://coverage.readthedocs.io/)
  1. Wrote initial unit-tests *for the CLI only at the moment* using `click.testing`
  1. Set up automation for tests using [`nox`](https://nox.thea.codes/), ensuring it uses pinned dependencies as declared in the `poetry.lock` to ensure testing amongst collaborates is identical. Note the `install_with_constraints()` wrapper in the `noxfile.py` as per [Claudio Jolowicz's suggestion](https://cjolowicz.github.io/posts/hypermodern-python-03-linting/#managing-dependencies-in-nox-sessions-with-poetry). `Nox` is now set up to test the script in `--no-dev` mode to reflect truer use, and not install all `dev` dependencies in each virtual environment.
  1. Expanded `nox` to run linting using [`flake8`](http://flake8.pycqa.org/), code formatting using [`black`](https://github.com/psf/black) and basic safety checks using [`safety`](https://github.com/pyupio/safety). Currently, the setup includes `flake8`, `flake8-bandit`, `flake8-black`, `flake8-bugbear`, and `flake8-import-order`. 
  1. Set up pre-commit hooks, and additionally runs `black` and `flake8` per each commit. This currently only checks changed files.
 

## Changed files

  1. `pyproject.toml` now includes `dev` dependencies for pytest (unit testing), coverage (unit test code coverage), black (auto code formatting), flake8 (linting and code-ordering)
  1. `noxfile.py` contains the setup for running `nox`
  1.  `tests/test_cli.py` contains test cases for the CLI interface. Test can be expanded on by adding files in the same folder, such as a `tests/test_module.py` file.
  1. `.flake8` contains settings for linting,
  1.  `.pre-commit-config.yaml` contains setttings for the pre-commit hooks. Changes to this file need to be installed via `pre-commit install` in the root folder. 
  1. Running linting resulted in changes to `csv2docx.py`, albeit none introduced changes to functionality.
  1. I've added `tests/data/example_missing_column.csv` to test our module catching a missing column in a `csv` file.

### Before
  1. Install `nox` on your machine globally, e.g. `pip install --user --upgrade nox
`. If running `nox` results in a `command not found`, ensure the path to the install is available to your console. E.g. I had to add `export PATH=/Users/<username>/.local/bin:$PATH` to my `~/.zshrc` file (I am using [OhMyZ](https://ohmyz.sh/).
  1. Install `pre-commit` on your machine globally, e.g. `pip install --user --upgrade pre-commit`
  1. Run `pre-commit install` to set up pre-commit 
  1. Run `poetry install` and `poetry shell` to update the dependencies (e.g. for `pytest`)

### Functionality
Nothing has changed to the module's functionality, so it should still work with:
- `poetry run convert -t tests/data/example.docx -c tests/data/example.csv -n NAME`
(perhaps run `poetry install` and `poetry shell` beforehand)

### Test suite
The following commands allow you to use the current setup:
- `poetry run pytest` to only run tests using `pytest`
- `nox -r` to run linting, safety and the pytests
- `nox -rs black` to perform code formatting according to `black` (this will also be done for each commit with `pre-commit`
- `pre-commit run --all-files` to run the `pre-commit` hooked actions without a commit (which will be executed for each `git commit`)

## Next Steps
- Some linting packages have not been added (yet), such as the suggested `flake8-builtins`, `flake8-quotes`, `flake8-docstrings` in #10. These can be added still. 
- Type annotations as per issue #11, e.g. following [this](https://cjolowicz.github.io/posts/hypermodern-python-04-typing/) guide.
- Expand testing, such as the character tests as mentioned in #23 
- Expand testing to validate document output, for example, whether .docx have been created (including edge cases).
- All guides strongly suggest to set up the package using a `/src/` folder - which we have discussed before. Potentially worth revisiting?

# Review
Please review @jawrainey. In particular, whether adding the dev dependencies was appropriate (reflected in `poetry.lock`), and if you'd agree with this initials setup and approach.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -14,10 +14,31 @@ docx-mailmerge = "^0.5.0"
 click = "^7.1.2"
 
 [tool.poetry.dev-dependencies]
```
[the remark this answers is not in the document]
Comment (jawrainey) [reply pyproject.toml:16]: Note: for `nox` to work when installed as dev dependency you must invoke it via poetry, e.g. `poetry run nox`

[the remark this answers is not in the document]
Comment (jawrainey) [reply pyproject.toml:16]: **Note:** `poetry run pre-commit install` in poetry's [contributing.md](https://github.com/python-poetry/poetry/blob/master/CONTRIBUTING.md), which is to ensure it's invoked within poetry's virtual environment rather than user level.

File: tests/test_cli.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    result = runner.invoke(
+        cli.main,
+        [
+            "--template",
+            "tests/data/example.docx",
+            "--data",
+            "tests/data/example.csv",
+            "--name",
+            "NAME",
+        ],
```
[the remark this answers is not in the document]
Comment (jawrainey) [reply tests/test_cli.py:28]: __Note:__ I have not tested the above so am uncertain if it'll work, but it gives an idea of what regarding using a fixture to reduce code duplication 👍

[the remark this answers is not in the document]
Comment (jawrainey) [reply tests/test_cli.py:28]: Perhaps an alternative (imho cleaner) approach to the above would be using [`parametrize`](https://docs.pytest.org/en/latest/parametrize.html). Open to debate 👍

File: tests/test_cli.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    )
+    assert result.exit_code == 0
+
+    # remove standard output files
+    for f in default_outpath.glob("*.docx"):
+        try:
+            f.unlink()
+        except OSError as e:
+            print("Error: %s : %s" % (f, e.strerror))
+
```
Comment (jawrainey) [inline tests/test_cli.py:38]: Couple of points:

- Use an f-string instead of string formatting (cleaner);
- Add `default_outpath.rmdir()` after the for loop to delete the `output` dir
- Remove the print statement and raise an assert instead since running pytests won't output the print statement, but an assert will.

This also raises an interesting question worth discussing around writing to/from files in tests, although arguments against this tend to be for projects that have many tests and therefore writing to disk slows down running tests [see here](https://endler.dev/2018/go-io-testing/). This is unlikely to impact us 👍 

The alternative is creating temporary files (as you've done below) or using _mocking_ with [unittest.mock](https://docs.python.org/3/library/unittest.mock.html). I think our current approach (reading/writing files) is fine (more functional testing than unit though?) and no need to use mocks as it'll introduce complexity that's not worth the effort imho.

**Comment by davidverweij** [comment]
Thanks @jawrainey! I'll go through the requested changes now and will update the pull request.

File: csv2docx/csv2docx.py [hunk 4]
```diff
@@ -1,7 +1,13 @@
...
 import csv
-from mailmerge import MailMerge
 from pathlib import Path
 
+# third-party packages
+from mailmerge import MailMerge
+
+# local packages
+# -- none
+
```
[the remark this answers is not in the document]
Comment (davidverweij) [reply csv2docx/csv2docx.py:10]: Apologies, I've added these comments myself as a reminder, but yes - I'll remove these.

File: pyproject.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidverweij) [reply pyproject.toml:16]: Good point, I followed the guide which indicated it would be good to have these as global packages. For my understanding, I had to do `poetry install`, `poetry shell` and then `pre-commit install` in order for it to set up the pre-commit hooks (I could then exit the shell and it would work). Am I missing a shortcut?

[the remark this answers is not in the document]
Comment (davidverweij) [reply pyproject.toml:16]: Woops - Now I get your comment about `poetry run pre-commit install`. Got it!

File: tests/test_cli.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            "tests/data/example.docx",
+            "--data",
+            "tests/data/example.csv",
+            "--name",
+            "NAME",
+            "--path",
+            tmpdir,
+        ],
+    )
+    assert result_dir.exit_code == 0
```
[the remark this answers is not in the document]
Comment (davidverweij) [reply tests/test_cli.py:89]: About (i), I tried adding `flake8-aaa` to the linter, but I keep getting an error 
> ERROR: In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    flake8-aaa from https://files.pythonhosted.org/packages/4d/52/8b023e4aa6d5f2bf668fb2c3bb3a1869a1eecda06ad345a69c8c1e14a8fd/flake8_aaa-0.9.0-py3-none-any.whl#sha256=828436c9746f912e5351c13846fd56071d4114ec91d93455c1424397eb35a8b6
And I don't see `flake8-aaa` being added to the poetry.lock file as a dev dependency (which should be giving it a pinned version). 

Before, these versions were pinned down when running the linter through `nox`, but I guess because I followed the guide to add the wrapper `install_with_constraints()` we need to explicitly install it as a dev dependency to pin down a version in the `poetry.lock` file. After `poetry add flake8-aaa --dev` it works nicely. Let me know if I missed something here.

File: tests/test_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jawrainey) [reply tests/test_cli.py:28]: re:parametrizing -- I looked into this and it is not as simple to implement this for the CLI as [`pytest.mark.parametrize`](https://docs.pytest.org/en/latest/parametrize.html) does not accept fixtures. See [this PyCon2020 video for an overview parameterized testing](https://us.pycon.org/2020/schedule/presentation/172/). As an example, here's what it might look like for testing the library under successful conditions:

```python

def library_success_loads():
    return  test_cases = [                                                                 
      ("tests/data/example.csv", "tests/data/example.docx", "NAME", "output", ";"),
      ("tests/data/example.csv", "tests/data/example.docx", "NAME", "nested/folder", ";"),
]                                                                              

@pytest.mark.parametrize("data, template, name, path, delimiter", test_cases)  
def test_library_convert_success(data, template, name, path, delimiter):               
     result = c2d.convert(data, template, name, path, delimiter)      
     # Note: not sure what to assert as convert success case returns None.     
     # which makes this comparison feel on pythonic  
     assert result  == None
```

File: tests/test_cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jawrainey) [reply tests/test_cli.py:89]: @davidverweij  -- adding it as a `--dev` and updating `def lint(session`) in `noxfile.py` to use the same plugin is the correct approach for installing and using the plugin 👍

File: tests/test_cli.py (hunk 3 above)
Comment (davidverweij) [reply tests/test_cli.py:38]: Got it, I've rewritten it as a f-string. It was a shameless act of copy-paste from an online solution.
I originally tested with `default_outpath.rmdir()` but it would error due to a non-empty folder (it contained a `.DS_Store`), but I've adjusted the code to completely empty the file.

Regarding the discussion point, the cleanup is only needed for the standard output, as the `tmp_path` and `tmpdir` are automatically cleaned  - they are kept for a t most 3 tests invocations. see [here](https://docs.pytest.org/en/latest/tmpdir.html).   So I do think we might not need the fixture. 

Regarding the unittests vs functional tests, yes, I would like to test whether the output is correct (e.g. the nameing of the .docx) but didn't got to it, so that would be more of a unittest don't you think?

**Comment by jawrainey** [comment]
> All guides strongly suggest to set up the package using a /src/ folder - which we have discussed before. Potentially worth revisiting?

@davidverweij -- could you also solve this? You'll need to:

1. Create `/src/` and move `/csv2docx` into it.
2. Update [code coverage config in pyproject](https://github.com/davidverweij/csv2docx/blob/unittests/pyproject.toml#L31-L32) to point to `"src"` instead of `"csv2docx"`. 

Remember to delete all `.pyc` or `__pycache__` files/folders before running poetry as it might not run correctly otherwise

File: tests/test_cli.py (hunk 3 above)
Comment (jawrainey) [reply tests/test_cli.py:38]: I was thinking that when we test the library by invoking `convert` directly that it would also create the `/output/<files>` and need the same cleanup. We can sort that out once we write tests for the library, but imho in a different issue since you have set the infrastructure in place here so we can progress type annotations etc alongside testing.

Yes, you're right about it being a unit test and great point about checking the output names of files. We should unit test `create_unique_name` specifically to verify naming, but of course we might want to test the default arguments/logic of the convert method also

Comment (davidverweij) [reply tests/test_cli.py]: Yes, could point about that. For now, I've set up the cleanup as fixture to also get rid of a very pressing `flake8-aaa` error ;-) 

I'll migrate these suggestions to another issue to expand on our tests.

File: tests/test_cli.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (davidverweij) [reply tests/test_cli.py]: Implemented the fixture, but moved the discussion of parametrizing to #23

**Comment by davidverweij** [comment]
>     1. Create `/src/` and move `/csv2docx` into it.
> 
>     2. Update [code coverage config in pyproject](https://github.com/davidverweij/csv2docx/blob/unittests/pyproject.toml#L31-L32) to point to `"src"` instead of `"csv2docx"`.

Done, but no need for the pointer to change, as the __init__.py is still in the csv2docx folder.

**Comment by jawrainey** [comment]
> > ```
> > 1. Create `/src/` and move `/csv2docx` into it.
> > 
> > 2. Update [code coverage config in pyproject](https://github.com/davidverweij/csv2docx/blob/unittests/pyproject.toml#L31-L32) to point to `"src"` instead of `"csv2docx"`.
> > ```
> 
> Done, but no need for the pointer to change, as the **init**.py is still in the csv2docx folder.

It is a bit confusing as no _init_ is required in the `src` folder, e.g. [see flask as an example](https://github.com/pallets/flask/tree/master/src). So our `convert` in `[tool.poetry.scripts]` likely won't need to change. _Why_ this works ... I am not certain

**Comment by davidverweij** [comment]
@jawrainey , I've hopefully addressed all requested changes - see my comments to your comments and in the [latest commit](https://github.com/davidverweij/csv2docx/pull/28/commits/9eaf2f3907cc2e934a9c0bddc24017d961290c8c). I've also moved the module to `/src/`, added a *contribute to code* and changed the pull request template both inspired on poetry's [contribution.md](https://github.com/python-poetry/poetry/blob/master/CONTRIBUTING.md). 

Please review and let me know if you'd want any changes!

**Comment by jawrainey** [comment]
I've pulled a clean copy of the repo and ran poetry as usual. The new infrastructure in nox for testing and linting, and pre-commit all work well. Great job on setting this up @davidverweij and writing the CLI unit tests. I especially liked the use of `usefixtures` and the tweak to the PR template was genius 🤯
